File: test-labsound.cpp
#include <memory>
#include <thread>

#include "test-labsound.hpp"
#include "LabSound/core/GainNode.h"

using lab::GainNode;
using lab::OscillatorNode;
using lab::OscillatorType;

int
main(int argc, char* argv[]) {
  AudioStreamConfig _inputConfig;
  AudioStreamConfig _outputConfig;
  auto config = GetDefaultAudioDeviceConfiguration(true);
  _inputConfig = config.first;
  _outputConfig = config.second;
  std::shared_ptr<lab::AudioDevice_RtAudio> device(
      new lab::AudioDevice_RtAudio(_inputConfig, _outputConfig));
  auto context = std::make_shared<lab::AudioContext>(false, true);
  auto destinationNode = std::make_shared<lab::AudioDestinationNode>(*context.get(), device);
  device->setDestinationNode(destinationNode);
  context->setDestinationNode(destinationNode);

  auto oscillator = std::make_shared<OscillatorNode>(*context.get());
  auto gain = std::make_shared<GainNode>(*context.get());
  gain->gain()->setValue(0.0625f);

  // osc -> gain -> destination
  context->connect(gain, oscillator, 0, 0);
  context->connect(context->destinationNode(), gain, 0, 0);

  oscillator->frequency()->setValue(440.f);
  oscillator->setType(OscillatorType::SINE);
  oscillator->start(0.0f);
  std::this_thread::sleep_for(std::chrono::seconds(1));
  return 0;
}

File: test-labsound.js
import { AudioDevice, AudioContext, AudioDestinationNode, OscillatorNode } from 'labsound';

function main() {
  const device = new AudioDevice({ deviceIndex: 0, desiredSampleRate: 44100, desiredChannels: 1 }, { deviceIndex: 0, desiredSampleRate: 44100, desiredChannels: 2 });

  const context = new AudioContext(false, true);

  const destinationNode = new AudioDestinationNode(context, device);

  device.destination = destinationNode;
  context.destination = destinationNode;

  const oscillator = new OscillatorNode(context);

  const gain = context.createGain();

  gain.gain = 0.0625;

  gain.connect(context.destination, 0, 0);

  oscillator.connect(gain, 0, 0);

  oscillator.frequency = 440.0;
  oscillator.type = 'Sine';
  oscillator.start(0);

  setTimeout(() => oscillator.stop(), 1000);
  setTimeout(() => {
    Object.assign(globalThis, { device, context, destinationNode, oscillator, gain });
    os.kill(os.getpid(), os.SIGUSR1);
  }, 1000);

  return 0;
}

main();
